File: selenium-e2e/src/test/java/dev/dsbon/realworld/pages/SideNav.java
package dev.dsbon.realworld.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

/** The left navigation drawer, present on every authenticated screen. */
public class SideNav extends BasePage {

  private static final By ROOT = testId("sidenav");
  private static final By USERNAME = testId("sidenav-username");
  private static final By FULL_NAME = testId("sidenav-user-full-name");
  private static final By BALANCE = testId("sidenav-user-balance");
  private static final By SIGN_OUT = testId("sidenav-signout");
  private static final By SETTINGS = testId("sidenav-user-settings");
  private static final By TOGGLE = testId("sidenav-toggle");

  public SideNav(WebDriver driver) {
    super(driver);
  }

  public SideNav awaitVisible() {
    visible(ROOT);
    return this;
  }

  /** Rendered as "@username". */
  public String username() {
    return textOf(USERNAME);
  }

  /** Rendered with an abbreviated surname, e.g. "Ted P". */
  public String fullName() {
    return textOf(FULL_NAME);
  }

  /** Rendered as a localized USD string, e.g. "$1,509.53". */
  public String balance() {
    return textOf(BALANCE);
  }

  public void signOut() {
    // Check the TARGET, not the hamburger.
    //
    // The first version asked "is the toggle displayed?" and clicked it if so.
    // On a desktop viewport the toggle IS displayed and the drawer is already
    // open, so that click CLOSED the drawer and made sign-out unreachable — a
    // self-inflicted timeout that only showed up in CI.
    //
    // Asking whether sign-out itself is reachable is correct at every size.
    if (driver.findElements(SIGN_OUT).stream().noneMatch(e -> e.isDisplayed())) {
      click(TOGGLE);
    }
    click(SIGN_OUT);
  }

  public void goToSettings() {
    click(SETTINGS);
  }
}

File: playwright-e2e/pages/SideNav.ts
import { expect, type Locator, type Page } from "@playwright/test";

/** The left navigation drawer, present on every authenticated screen. */
export class SideNav {
  readonly root: Locator;
  readonly username: Locator;
  readonly fullName: Locator;
  readonly balance: Locator;
  readonly signOut: Locator;
  readonly settings: Locator;
  readonly home: Locator;

  constructor(private readonly page: Page) {
    this.root = page.getByTestId("sidenav");
    this.username = page.getByTestId("sidenav-username");
    this.fullName = page.getByTestId("sidenav-user-full-name");
    this.balance = page.getByTestId("sidenav-user-balance");
    this.signOut = page.getByTestId("sidenav-signout");
    this.settings = page.getByTestId("sidenav-user-settings");
    this.home = page.getByTestId("sidenav-home");
  }

  async expectSignedInAs(username: string): Promise<void> {
    await expect(this.username).toContainText(`@${username}`);
  }

  /** The drawer abbreviates the surname to an initial, e.g. "Ted P". */
  async expectFullName(firstName: string, lastName: string): Promise<void> {
    await expect(this.fullName).toContainText(firstName);
    await expect(this.fullName).toContainText(lastName.charAt(0));
  }

  /** Balances render as a localized USD string, e.g. "$1,509.53". */
  async expectFormattedBalance(): Promise<void> {
    await expect(this.balance).toHaveText(/^\$[\d,]+\.\d{2}$/);
  }

  async clickSignOut(): Promise<void> {
    // Check the TARGET, not the hamburger.
    //
    // The first version of this asked "is the toggle visible?" and clicked it if
    // so. On a desktop viewport the toggle is visible AND the drawer is already
    // open, so that click CLOSED the drawer and the sign-out item became
    // unclickable — a self-inflicted timeout that only appeared in CI.
    //
    // Asking whether sign-out itself is reachable is correct at every viewport.
    if (!(await this.signOut.isVisible())) {
      await this.page.getByTestId("sidenav-toggle").click();
    }
    await this.signOut.click();
  }
}
